src/MiniProfiler.Providers.StackExchange.Redis/RedisStorage.cs

```diff
@@ -0,0 +1,263 @@
+        /// <param name="orderBy">(Optional) The order to fetch profiler IDs in.</param>
+        public async Task<IEnumerable<Guid>> ListAsync(
+            int maxResults,
+            DateTime? start = null,
+            DateTime? finish = null,
+            ListResultsOrder orderBy = ListResultsOrder.Descending)
+        {
+            // TODO: start/finish not yet supported - will simply return all results
+
+            IEnumerable<RedisValue> ids;
+            switch (orderBy)
+            {
+                case ListResultsOrder.Ascending:
+                    ids = (await _database.ListRangeAsync(ProfilerResultListKey, maxResults + 1, -1).ConfigureAwait(false)).Reverse();
+                    break;
+                case ListResultsOrder.Descending:
+                default:
+                    ids = await _database.ListRangeAsync(ProfilerResultListKey, 0, maxResults - 1).ConfigureAwait(false);
+                    break;
+            }
+            return ids.Select(x => Guid.Parse(x));
+        }
+
+        /// <summary>
+        /// Asynchronously stores <paramref name="profiler"/> under its <see cref="MiniProfiler.Id"/>.
+        /// </summary>
+        /// <param name="profiler">The <see cref="MiniProfiler"/> to save.</param>
+        /// <remarks>
+        /// Should also ensure the profiler is stored as being un-viewed by its profiling <see cref="MiniProfiler.User"/>.
+        /// </remarks>
+        public async Task SaveAsync(MiniProfiler profiler)
+        {
+            var id = profiler.Id.ToString();
+            RedisKey key = ProfilerResultKeyPrefix.Append(id);
+            RedisValue value = profiler.ToJson();
+
+            await _database.StringSetAsync(key, value, expiry: CacheDuration).ConfigureAwait(false);
+
+            await _database.ListLeftPushAsync(ProfilerResultListKey, id).ConfigureAwait(false);
+            await _database.ListTrimAsync(ProfilerResultListKey, 0, ResultListMaxLength - 1).ConfigureAwait(false);
+            await _database.KeyExpireAsync(ProfilerResultListKey, CacheDuration).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Asynchronously returns a <see cref="MiniProfiler"/> from storage based on <paramref name="id"/>,
+        /// which should map to <see cref="MiniProfiler.Id"/>.
+        /// </summary>
+        /// <param name="id">The profiler ID to load.</param>
+        /// <returns>The loaded <see cref="MiniProfiler"/>.</returns>
+        /// <remarks>
+        /// Should also update that the resulting profiler has been marked as viewed by its
+        /// profiling <see cref="MiniProfiler.User"/>.
+        /// </remarks>
+        public async Task<MiniProfiler> LoadAsync(Guid id)
+        {
+            RedisKey key = ProfilerResultKeyPrefix.Append(id.ToString());
+            RedisValue value = await _database.StringGetAsync(key).ConfigureAwait(false); ;
+            return MiniProfiler.FromJson(value);
```

Comment:
I'm not necessarily opposed to taking on protobuf-net as a dependency. But maybe instead we could allow users to customize the serialization - either an interface or just a pair of delegates - to avoid the hard dependency. I assume it could a pretty straightforward call.

e.g. `RedisValue Encode(MiniProfiler profiler);`